lxd/instances_post.go

```diff
@@ -880,65 +874,11 @@ func instancesPost(d *Daemon, r *http.Request) response.Response {
 			return err
 		}
 
-		var allMembers []db.NodeInfo
-
 		if clustered && !clusterNotification {
-			clusterGroupsAllowed = shared.SplitNTrimSpace(targetProject.Config["restricted.cluster.groups"], ",", -1, true)
-
-			// Check manual cluster member targeting restrictions.
-			err = project.CheckClusterTargetRestriction(r, targetProject, target)
+			targetMemberInfo, err = project.CheckTarget(ctx, r, tx, targetProject, target)
```

Comment:
This change caught my eye. Not because it doesn't clean up the code and improve the readability, it certainly does.

However, what caught by eye was the change of scope of allMembers var definition above, and the moving of the tx.GetNodes call to further down in the function.

On closer inspection, the reason you've been able to do this is because the `CheckTarget` function calls `CheckTargetMember` which in turn calls `tx.GetNodeByName`. 

This may seem harmless, but its important to remember that while dqlite is based upon sqlite, it does not have the same access characteristics. 

The key things to remember with dqlite are:

1. All queries are sent to the leader (which may not be this cluster member at the time) so can incur network latency.
2. Transactions can also incur network latency due to additional round trips (in order to start and commit them), so try and keep them to a minimal, which means doing as many queries as you can inside fewer transactions. Whilst also considering point 3...
3. Locking is at the database level (so don't keep transactions open longer than necessary).

So that extra call to `tx.GetNodeByName` is performing 4 extra queries (as that function loads the node info, node group info, node role info and node config).

Also this is all happening within the wider context of an instance creation request, which is something that occurs often and we should try and keep fast and lean (as opposed to some operation that happens only once in a blue moon).

So I think we should keep the tx.GetNodes where it is, and instead have `project.CheckTarget` accept a list of potential candidates which we can pass the list of nodes we already retrieved into.

Reply:
`tx.GetNodeByName` is now called once and the filtering is performed later on the already existing list of nodes.